src/java.base/share/classes/java/lang/foreign/Arena.java

```diff
@@ -0,0 +1,121 @@
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package java.lang.foreign;
+
+import jdk.internal.foreign.ArenaImpl;
+import jdk.internal.foreign.MemorySessionImpl;
+import jdk.internal.javac.PreviewFeature;
+
+/**
+ * An arena allocates and manages the lifecycle of native segments.
+ * <p>
+ * An arena is a {@linkplain AutoCloseable closeable} segment allocator that is associated with a {@link #session() memory session}.
+ * This session is created with the arena, and is closed when the arena is {@linkplain #close() closed}.
+ * Furthermore, all the native segments {@linkplain #allocate(long, long) allocated} by the arena are associated
+ * with that session.
+ * <p>
+ * The <a href="MemorySession.html#thread-confinement">confinement properties</a> of the session associated with an
+ * arena are determined by the factory used to create the arena. For instance, an arena created with {@link #openConfined()}
+ * is associated with a <em>confined</em> memory session. Conversely, an arena created with {@link #openShared()} is
+ * associated with a <em>shared</em> memory session.
+ * <p>
+ * An arena is extremely useful when interacting with foreign code, as shown below:
+ *
+ * {@snippet lang = java:
+ * try (Arena arena = Arena.openConfined()) {
+ *     MemorySegment nativeArray = arena.allocateArray(ValueLayout.JAVA_INT, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9);
+ *     MemorySegment nativeString = arena.allocateUtf8String("Hello!");
+ *     MemorySegment upcallStub = linker.upcallStub(handle, desc, arena.session());
+ *     ...
+ * } // memory released here
+ *}
+ *
+ * @since 20
+ */
+@PreviewFeature(feature=PreviewFeature.Feature.FOREIGN)
+public sealed interface Arena extends SegmentAllocator, AutoCloseable permits ArenaImpl { //@@@: This can also be non-sealed!
+
+    /**
+     * Creates a native memory segment with the given size (in bytes), alignment constraint (in bytes) associated with
+     * this memory session. The {@link MemorySegment#address()} of the returned memory segment is the starting address of
+     * the newly allocated off-heap memory region backing the segment. Moreover, the {@linkplain MemorySegment#address() address}
+     * of the returned segment will be aligned according the provided alignment constraint.
+     * <p>
+     * Clients are responsible for ensuring that this arena is closed when the segments returned by this method are no
+     * longer in use. Failure to do so will result in off-heap memory leaks.
+     * <p>
+     * This is equivalent to the following code:
```

Comment:
If we would allow creating custom arenas than this statement may not be true (ie. of pooling arena)